### ISSUE TITLE
请问一下`dropout`的单测是如何判定计算结果正确性的

### ISSUE THREAD

**Issue by houj04** [PaddlePaddle/Paddle#62642]
通常算子的单测，是GPU跑一遍，CPU跑一遍，然后比较计算结果的diff。但是我看`dropout`这个算子，它的CPU实现
`paddle/phi/kernels/cpu/dropout_kernel.cc`
和GPU实现
`paddle/phi/kernels/gpu/dropout_kernel.cu`以及`paddle/phi/kernels/funcs/dropout_impl.cu.h`
从随机数生成的方式上就不一样。CPU是用的`std::shared_ptr<std::mt19937_64>`，而GPU用的是`curandStatePhilox4_32_10_t`，这两个随机数生成器，即使喂同样的随机种子，出来的结果也不一样吧。
想问的是，对于这种随机生成的算子，单测是如何执行的呢？换句话说，假设我尝试“优化”或者“修改”这个算子，如何知道我“没有改坏”呢？

补充：我注意到`gaussian`随机数生成的单测`test/legacy_test/test_gaussian_random_op.py`里面有计算hist的步骤，看起来是计算生成的分布的hist是否符合预期。但是`dropout`的单测里面好像没有类似的代码。

**Comment by wanghuancoder** [comment]
这个问题请 @ceci3 帮忙解答一下。

**Comment by ceci3** [comment]
固定seed去做判断，看结果能不能保持一致呢

**Comment by houj04** [comment]
> 固定seed去做判断，看结果能不能保持一致呢

啊，我想问的是，现有的单测，是如何判定计算结果正确性的。在单测代码里面，好像没找到这种判断一致的地方。
另，我总觉得在CPU上和GPU上要想保证随机生成的数都一样，实在是太难了。

**Comment by houj04** [comment]
补充一下背景：我在写XPU下面的`dropout`算子，写完了能跑不挂，但是不知道如何验证它的正确性。例如，当`dropout_prob`为0.1的时候，按照设计应该是干掉10%的数据，剩下90%保留。但是我有可能有bug，写成了干掉90%的数据而只保留10%。或者，它明明应该是“随机干掉”，但是我有可能有bug，实现成了“固定干掉开头的10%”。
我想用单测的方式来提升代码质量，但是看了一下GPU的单测里面，没有这种明确比较结果的地方，是不是放到了其它什么地方？

然后，我试了一下，用同样的随机种子，CPU和GPU生成的结果是不一样的。代码如下：

```python
import paddle

paddle.set_device('gpu')
paddle.seed(2023)
x = paddle.to_tensor([[1,2,3], [4,5,6]]).astype(paddle.float32)
y_train = paddle.nn.functional.dropout(x, 0.5)
print(x)
print(y_train)

paddle.set_device('cpu')
paddle.seed(2023)
x = paddle.to_tensor([[1,2,3], [4,5,6]]).astype(paddle.float32)
y_train = paddle.nn.functional.dropout(x, 0.5)
print(x)
print(y_train)
```

运行结果如下：

> Tensor(shape=[2, 3], dtype=float32, place=Place(gpu:0), stop_gradient=True,
>        [[1., 2., 3.],
>         [4., 5., 6.]])
> Tensor(shape=[2, 3], dtype=float32, place=Place(gpu:0), stop_gradient=True,
>        [[0., 0., 6.],
>         [0., 0., 0.]])
> Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
>        [[1., 2., 3.],
>         [4., 5., 6.]])
> Tensor(shape=[2, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
>        [[2., 4., 0.],
>         [8., 0., 0.]])

**Comment by zhwesky2010** [comment]
@houj04 正如你上面提到的CPU是用的`std::shared_ptr<std::mt19937_64>`，而GPU用的是`curandStatePhilox4_32_10_t`，这两个随机数生成算法，运行的结果是无法一致的，因此CPU与GPU不同是符合预期的。

目前的 `test_dropout.py` 里，是绕过了数值比较，大部分是在 `p=0.0/1.0` 这两种情况下来对比确定性结果，对于 `p!=0.0/1.0` 情况下，设置了`is_test=True`，此时只会跑推理情况`out=x*(1-p)`，从而来对比确定性的结果。

只有这个case里是进行了具体的值来比较：https://github.com/PaddlePaddle/Paddle/blob/develop/test/legacy_test/test_dropout_op.py#L1435 ，参考较少，如果你需要比对值的话，建议可以从**统计非0数量**的角度来开展。

**Comment by houj04** [comment]
了解，多谢解答！
